PLANET-7489 Migrate Campaign Covers to P4 Columns block

sagarsdeshmukh opened the pull request:
Add a migration script which transform campaign covers block into planet4 columns block

Ref. https://jira.greenpeace.org/browse/PLANET-7489

**Testing:**
- On local dev env, run `npx wp-env run cli wp p4-run-activator` to run the migration script.

[Ref.  ](https://app.gitbook.com/o/-LMm4Q4AuKcwl38JYrxF/s/-M15KrJzoMvhbv4NcO9o/tech/data-migrations)


### Post deploy script execution on test instance -
https://app.circleci.com/pipelines/github/greenpeace/planet4-test-mars/734/workflows/e33ae2e2-ee71-473b-a1b2-b3c3e8fcd1e9/jobs/2351/parallel-runs/0/steps/0-108

```
Running the script: 01-run-p4-activator.sh

Run P4_Activator...
Parsing post 810
Migration successful
Parsing post 69
Migration successful
Parsing post 60
Migration successful
```

**Comparison of same posts on [mars-test](https://www-dev.greenpeace.org/test-mars/wp-admin) (with migration) & [titan-test](https://www-dev.greenpeace.org/test-titan/wp-admin) (without migration) -**

1. Please check the "Covers block" > "Campaign covers style" on page

https://www-dev.greenpeace.org/test-mars/testing-all-blocks-evergreen-template/

Similar page on test-titan instance
https://www-dev.greenpeace.org/test-titan/testing-all-blocks-evergreen-template/#covers-block-2

2.
https://www-dev.greenpeace.org/test-mars/get-informed/people/#:~:text=at%20eleifend%20ullamcorper.-,Our%20campaigns,-%23Consumption

https://www-dev.greenpeace.org/test-titan/get-informed/people/#:~:text=at%20eleifend%20ullamcorper.-,Our%20campaigns,-%23Climate

3.
https://www-dev.greenpeace.org/test-mars/get-informed/energy/

https://www-dev.greenpeace.org/test-titan/get-informed/energy/

mleray commented:
@sagarsdeshmukh I don't know if you saw that the PHP tests failed, should be quite straightforward to fix 🙂

planet-4 commented:
### Test instance is ready :rocket:

:new_moon: [mars](https://www-dev.greenpeace.org/test-mars) | [admin](https://www-dev.greenpeace.org/test-mars/wp-admin/) | [blocks report](https://www-dev.greenpeace.org/test-mars/wp-admin/?page=plugin_blocks_report) | [CircleCI](https://app.circleci.com/pipelines/github/greenpeace/planet4-test-mars) | [composer-local.json](https://github.com/greenpeace/planet4-test-mars/blob/main/composer-local.json)

:watch: 2024.08.20 07:26:56

mardelnet reviewed (commented):
Good job, @sagarsdeshmukh 
I see visual differences between the Campaign Covers and the P4 Columns block, is that expected, or should they look exactly the same?

![Screenshot from 2024-08-21 09-23-39](https://github.com/user-attachments/assets/8d8648fd-a8f1-4b93-9ca8-c43845c6042b)

sagarsdeshmukh commented:
> Good job, @sagarsdeshmukh I see visual differences between the Campaign Covers and the P4 Columns block, is that expected, or should they look exactly the same?
> 
> ![Screenshot from 2024-08-21 09-23-39](https://private-user-images.githubusercontent.com/59029273/359793661-8d8648fd-a8f1-4b93-9ca8-c43845c6042b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQyMjU4MDMsIm5iZiI6MTcyNDIyNTUwMywicGF0aCI6Ii81OTAyOTI3My8zNTk3OTM2NjEtOGQ4NjQ4ZmQtYThmMS00YjkzLTljYTgtYzQzODQ1YzYwNDJiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIxVDA3MzE0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZlNzdjZWY0M2I0MWZiMTY0ZDQyYjZhYzJlNTZkMDhhZDQ3MzZiMzc5MzIwMjNhOWYxODI4YWI5YzNiNTY0MDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.bH_3Vb2ynlZ9myncU3zC8g9s9bE7HXdyh8pciIlf8jo)

Thanks, Pedro. Based on the ticket details, I don't think they will look exactly the same. We're trying to transform the existing campaign covers into P4 columns, so they will resemble the P4 columns block but with content from the campaign covers.

mardelnet reviewed (commented):
Hey @sagarsdeshmukh, just to be sure: we're only migrating the Cover blocks of type Campaign, right? Aren't we migrating the Cover blocks of type Take Action and Content at this point, as well?

mardelnet reviewed (commented):
After running the migration script on my local environment, the Campaign Cover blocks were successfully turned into P4 blocks, but the Content Cover and Take Action Cover blocks were removed. Is that something expected?

mardelnet reviewed (commented):
Hey @sagarsdeshmukh 

1. I just made a test and ran the migrator, and only the last one was transformed into columns on a page with multiple Campaign Cover blocks. Would you mind checking that?
2. I think I found a way to transform blocks using the WP_Block_Parser class and avoiding the need to find and match code using regex. You can check it [here](https://github.com/greenpeace/planet4-master-theme/compare/main...PLANET-7521_migrate-media-block-to-audio-video-blocks#diff-053e954328867930e539fca66766722a453a9f544400e19f85cc1bacb0b2680fR52).

sagarsdeshmukh commented:
> Hey @sagarsdeshmukh, just to be sure: we're only migrating the Cover blocks of type Campaign, right? Aren't we migrating the Cover blocks of type Take Action and Content at this point, as well?

Hi @mardelnet , The covers Take Action and Content styles are planned to migrate in separate tickets.

sagarsdeshmukh commented:
> After running the migration script on my local environment, the Campaign Cover blocks were successfully turned into P4 blocks, but the Content Cover and Take Action Cover blocks were removed. Is that something expected?

It's not expected, I updated the PR. Could you please double-check now?

mardelnet commented on src/Migrations/M032MigrateCampaignCoversToP4ColumnsBlock.php:
```diff
@@ -0,0 +1,203 @@
+ */
+class M032MigrateCampaignCoversToP4ColumnsBlock extends MigrationScript
+{
+    private const BLOCK_NAME = 'planet4-blocks/covers';
+    private const POST_TYPES = [ 'page', 'post', 'action', 'campaign' ];
+    private const COVER_TYPE = 'campaign';
+    private const OLD_COVER_TYPES = [
+        '1' => 'take-action',
+        '2' => 'campaign',
+        '3' => 'content',
+    ];
```
Some of these constants can be reused in other files (for example, [here](https://github.com/greenpeace/planet4-master-theme/pull/2363/files#diff-882ae83a59536852c35295e4efeeaf80e277d1f512cb5c10d2143e9fbad46c76R17)).
Do you think it is worth moving them to a separate file to avoid duplicate content?

mardelnet commented on src/Migrations/M032MigrateCampaignCoversToP4ColumnsBlock.php:
```diff
@@ -0,0 +1,203 @@
+     * @param string $block_name - A block name.
+     * @return mixed - The posts using block or null if no posts are found.
+     */
+    private static function get_posts_using_block(string $block_name): mixed
```
This function is also used [here](https://github.com/greenpeace/planet4-master-theme/pull/2363/files#diff-882ae83a59536852c35295e4efeeaf80e277d1f512cb5c10d2143e9fbad46c76R88) and [here](https://github.com/greenpeace/planet4-master-theme/pull/2357/files#diff-053e954328867930e539fca66766722a453a9f544400e19f85cc1bacb0b2680fR110).
Do you think it is worth moving it to a separate file to avoid duplicate content?

mardelnet commented on src/Migrations/M032MigrateCampaignCoversToP4ColumnsBlock.php:
```diff
@@ -0,0 +1,203 @@
+                continue;
+            }
+
+            if ('title' === $key) {
```
Do we need the `if` and `else if` statements here?
Since we are only looking for the title, description, and tags maybe we could do something like this to improve the readability:

```php
private static function get_columns_block_attrs(array $block): array
    {
        $block_attrs = [];
        $block_attrs['columns_title'] = $block['attrs']['title'] ?? '';
        $block_attrs['columns_description'] = $block['attrs']['description'] ? : str_replace("\n", "\u003cbr\u003e", $block['attrs']['description']); '';
         ...
```

mardelnet commented on src/Migrations/M032MigrateCampaignCoversToP4ColumnsBlock.php:
```diff
@@ -0,0 +1,203 @@
+     */
+    public static function execute(MigrationRecord $record): void
+    {
+        $parser = new WP_Block_Parser();
```
Maybe this line could be moved above line 44 to only create a new instance of the `WP_Block_Parser`  if posts are using the Campaign Covers block.

mardelnet commented on src/Migrations/M032MigrateCampaignCoversToP4ColumnsBlock.php:
```diff
@@ -0,0 +1,203 @@
+    private const COVER_TYPE = 'campaign';
+    private const OLD_COVER_TYPES = [
+        '1' => 'take-action',
+        '2' => 'campaign',
```
Maybe, to avoid repetition:
```suggestion
        '2' => self::COVER_TYPE,
```

mardelnet commented on src/Migrations/M032MigrateCampaignCoversToP4ColumnsBlock.php:
```diff
@@ -0,0 +1,203 @@
+            $blocks = $parser->parse($post->post_content);
+
+            foreach ($blocks as &$block) {
+                // For older cover blocks, the cover type is empty, so use the default content cover type,hence skip it.
+                if (!isset($block['attrs']['cover_type'])) {
+                    continue;
+                }
+
+                // For old cover type(earlier it was numeric).
+                if (is_numeric($block['attrs']['cover_type'])) {
+                    $block['attrs']['cover_type'] = self::OLD_COVER_TYPES[ $block['attrs']['cover_type'] ];
+                }
+
+                // Skip other blocks & non campaign cover blocks.
+                if (
+                    ! isset($block['blockName']) || $block['blockName'] !== self::BLOCK_NAME ||
+                    $block['attrs']['cover_type'] !== self::COVER_TYPE
+                ) {
+                    continue;
+                }
```
I think these guard clauses should be re-ordered.
IMO, we should first check whether the block name is right, then check the cover type, and finally update the numeric value, something like:

```php
if (!isset($block['blockName']) || $block['blockName'] !== self::BLOCK_NAME) {
             continue;
}
if (!isset($block['attrs']['cover_type']) || $block['attrs']['cover_type'] !== self::COVER_TYPE) {
       continue;
}
if (is_numeric($block['attrs']['cover_type'])) {
         $block['attrs']['cover_type'] = self::OLD_COVER_TYPES[ $block['attrs']['cover_type'] ];
}
```

sagarsdeshmukh replied:
I agree on moving the Block name check to top, but cover type check need to keep below numeric cover type check, cause else we will skip all numeric cover type blocks. I adjusted code as below -

`
                // Skip non cover blocks.
                if (!isset($block['blockName']) || $block['blockName'] !== self::BLOCK_NAME) {
                    continue;
                }

                // For older cover blocks, the cover type is empty, so use the default content cover type, hence skip it.
                if (!isset($block['attrs']['cover_type'])) {
                    continue;
                }

                // For old cover type(earlier it was numeric).
                if (is_numeric($block['attrs']['cover_type'])) {
                    $block['attrs']['cover_type'] = self::OLD_COVER_TYPES[ $block['attrs']['cover_type'] ];
                }

                // Skip non campaign cover blocks.
                if ($block['attrs']['cover_type'] !== self::COVER_TYPE) {
                    continue;
                }
`

mardelnet commented on src/Migrations/M032MigrateCampaignCoversToP4ColumnsBlock.php:
```diff
@@ -0,0 +1,203 @@
+                    continue;
+                }
+
+                $block_attrs = array_merge(['columns_block_style' => 'image'], $block_attrs);
+                $block_attrs['className'] = 'is-style-image';
```
Should this be part of the `get_columns_block_attrs` function?

sagarsdeshmukh commented:
Hi @mardelnet , The PR is updated with reusable code and `PLANET-7521_migrate-media-block-to-audio-video-blocks` branch as a base branch. Please have a look.

mardelnet reviewed (commented):
Nice job, @sagarsdeshmukh 
I still see a couple of issues:
1. After running the migration script, Content-type cover blocks crash in the editor and the "Attempt block recovery" button appears. You can locally test it using the blocks code I added at the end of this message.
2. The line breaks of the cover block description are removed and replaced by the string `u003cbru003e` after the migration.

``` 
<!-- wp:planet4-blocks/covers {"cover_type":"take-action","title":"Take Action Cover Block","className":"is-style-take-action"} -->
<div class="wp-block-planet4-blocks-covers is-style-take-action"><div data-hydrate="planet4-blocks/covers" data-attributes="{&quot;cover_type&quot;:&quot;take-action&quot;,&quot;initialRowsLimit&quot;:1,&quot;title&quot;:&quot;Take Action Cover Block&quot;,&quot;description&quot;:&quot;&quot;,&quot;tags&quot;:[],&quot;post_types&quot;:[],&quot;posts&quot;:[],&quot;version&quot;:2,&quot;layout&quot;:&quot;grid&quot;,&quot;isExample&quot;:false,&quot;readMoreText&quot;:&quot;Load more&quot;,&quot;covers&quot;:[],&quot;className&quot;:&quot;is-style-take-action&quot;}"></div></div>
<!-- /wp:planet4-blocks/covers -->

<!-- wp:planet4-blocks/covers {"cover_type":"campaign","title":"Campaign Cover Block","tags":[8],"className":"is-style-campaign"} -->
<div class="wp-block-planet4-blocks-covers is-style-campaign"><div data-hydrate="planet4-blocks/covers" data-attributes="{&quot;cover_type&quot;:&quot;campaign&quot;,&quot;initialRowsLimit&quot;:1,&quot;title&quot;:&quot;Campaign Cover Block&quot;,&quot;description&quot;:&quot;&quot;,&quot;tags&quot;:[8],&quot;post_types&quot;:[],&quot;posts&quot;:[],&quot;version&quot;:2,&quot;layout&quot;:&quot;grid&quot;,&quot;isExample&quot;:false,&quot;readMoreText&quot;:&quot;Load more&quot;,&quot;covers&quot;:[],&quot;className&quot;:&quot;is-style-campaign&quot;}"></div></div>
<!-- /wp:planet4-blocks/covers -->

<!-- wp:planet4-blocks/covers {"cover_type":"campaign","title":"Campaign Cover Block 2","description":"Description of the campaign cover 2\u003cbr\u003eLayout: carousel\u003cbr\u003eTags: consumption + renewables","tags":[8,7],"layout":"carousel","className":"is-style-campaign"} -->
<div class="wp-block-planet4-blocks-covers is-style-campaign"><div data-hydrate="planet4-blocks/covers" data-attributes="{&quot;cover_type&quot;:&quot;campaign&quot;,&quot;initialRowsLimit&quot;:1,&quot;title&quot;:&quot;Campaign Cover Block 2&quot;,&quot;description&quot;:&quot;Description of the campaign cover 2&lt;br&gt;Layout: carousel&lt;br&gt;Tags: consumption + renewables&quot;,&quot;tags&quot;:[8,7],&quot;post_types&quot;:[],&quot;posts&quot;:[],&quot;version&quot;:2,&quot;layout&quot;:&quot;carousel&quot;,&quot;isExample&quot;:false,&quot;readMoreText&quot;:&quot;Load more&quot;,&quot;covers&quot;:[],&quot;className&quot;:&quot;is-style-campaign&quot;}"></div></div>
<!-- /wp:planet4-blocks/covers -->

<!-- wp:planet4-blocks/covers {"cover_type":"campaign","initialRowsLimit":2,"title":"Campaign Cover Block 3","description":"Description of the campaign cover 2\u003cbr\u003eLayout: grid 2 rows\u003cbr\u003eTags: consumption + renewables + oceans + climate","tags":[8,7,6,26],"className":"is-style-campaign"} -->
<div class="wp-block-planet4-blocks-covers is-style-campaign"><div data-hydrate="planet4-blocks/covers" data-attributes="{&quot;cover_type&quot;:&quot;campaign&quot;,&quot;initialRowsLimit&quot;:2,&quot;title&quot;:&quot;Campaign Cover Block 3&quot;,&quot;description&quot;:&quot;Description of the campaign cover 2&lt;br&gt;Layout: grid 2 rows&lt;br&gt;Tags: consumption + renewables + oceans + climate&quot;,&quot;tags&quot;:[8,7,6,26],&quot;post_types&quot;:[],&quot;posts&quot;:[],&quot;version&quot;:2,&quot;layout&quot;:&quot;grid&quot;,&quot;isExample&quot;:false,&quot;readMoreText&quot;:&quot;Load more&quot;,&quot;covers&quot;:[],&quot;className&quot;:&quot;is-style-campaign&quot;}"></div></div>
<!-- /wp:planet4-blocks/covers -->

<!-- wp:planet4-blocks/covers {"title":"Content Cover Block 1","description":"Layout: carousel\u003cbr\u003eTags: consumption, renewables\u003cbr\u003ePost type: press release","tags":[8,7],"post_types":[14],"layout":"carousel","className":"is-style-content"} -->
<div class="wp-block-planet4-blocks-covers is-style-content"><div data-hydrate="planet4-blocks/covers" data-attributes="{&quot;cover_type&quot;:&quot;content&quot;,&quot;initialRowsLimit&quot;:1,&quot;title&quot;:&quot;Content Cover Block 1&quot;,&quot;description&quot;:&quot;Layout: carousel&lt;br&gt;Tags: consumption, renewables&lt;br&gt;Post type: press release&quot;,&quot;tags&quot;:[8,7],&quot;post_types&quot;:[14],&quot;posts&quot;:[],&quot;version&quot;:2,&quot;layout&quot;:&quot;carousel&quot;,&quot;isExample&quot;:false,&quot;readMoreText&quot;:&quot;Load more&quot;,&quot;covers&quot;:[],&quot;className&quot;:&quot;is-style-content&quot;}"></div></div>
<!-- /wp:planet4-blocks/covers -->

<!-- wp:planet4-blocks/covers {"title":"Content Cover Block 2","description":"Layout: grid\u003cbr\u003eTags: -\u003cbr\u003ePost type: story, press release","post_types":[16,14],"className":"is-style-content"} -->
<div class="wp-block-planet4-blocks-covers is-style-content"><div data-hydrate="planet4-blocks/covers" data-attributes="{&quot;cover_type&quot;:&quot;content&quot;,&quot;initialRowsLimit&quot;:1,&quot;title&quot;:&quot;Content Cover Block 2&quot;,&quot;description&quot;:&quot;Layout: grid&lt;br&gt;Tags: -&lt;br&gt;Post type: story, press release&quot;,&quot;tags&quot;:[],&quot;post_types&quot;:[16,14],&quot;posts&quot;:[],&quot;version&quot;:2,&quot;layout&quot;:&quot;grid&quot;,&quot;isExample&quot;:false,&quot;readMoreText&quot;:&quot;Load more&quot;,&quot;covers&quot;:[],&quot;className&quot;:&quot;is-style-content&quot;}"></div></div>
<!-- /wp:planet4-blocks/covers -->

<!-- wp:planet4-blocks/covers {"title":"Content Cover Block 3","description":"Layout: grid\u003cbr\u003eTags: consumption, renewables\u003cbr\u003ePost type: -","tags":[8,7],"className":"is-style-content"} -->
<div class="wp-block-planet4-blocks-covers is-style-content"><div data-hydrate="planet4-blocks/covers" data-attributes="{&quot;cover_type&quot;:&quot;content&quot;,&quot;initialRowsLimit&quot;:1,&quot;title&quot;:&quot;Content Cover Block 3&quot;,&quot;description&quot;:&quot;Layout: grid&lt;br&gt;Tags: consumption, renewables&lt;br&gt;Post type: -&quot;,&quot;tags&quot;:[8,7],&quot;post_types&quot;:[],&quot;posts&quot;:[],&quot;version&quot;:2,&quot;layout&quot;:&quot;grid&quot;,&quot;isExample&quot;:false,&quot;readMoreText&quot;:&quot;Load more&quot;,&quot;covers&quot;:[],&quot;className&quot;:&quot;is-style-content&quot;}"></div></div>
<!-- /wp:planet4-blocks/covers -->
```

mardelnet commented on src/Migrations/M032MigrateCampaignCoversToP4ColumnsBlock.php:
```diff
@@ -0,0 +1,160 @@
+                'post_content' => $new_content,
+            );
+
+            try {
+                // Update the post with the replaced blocks.
+                wp_update_post($post_update);
+                $result = true;
+            } catch (\Throwable $e) {
+                echo 'Error on post ', $post->ID, "\n";
+                echo $e->getMessage(), "\n";
+            }
```
Hey @sagarsdeshmukh 
Shouldn't we wrap the whole logic of this function on `try/catch` block instead of only this part?

sagarsdeshmukh commented:
> Nice job, @sagarsdeshmukh I still see a couple of issues:
> 
> 1. After running the migration script, Content-type cover blocks crash in the editor and the "Attempt block recovery" button appears. You can locally test it using the blocks code I added at the end of this message.
> 2. The line breaks of the cover block description are removed and replaced by the string `u003cbru003e` after the migration.
> 
> ```
> <!-- wp:planet4-blocks/covers {"cover_type":"take-action","title":"Take Action Cover Block","className":"is-style-take-action"} -->
> <div class="wp-block-planet4-blocks-covers is-style-take-action"><div data-hydrate="planet4-blocks/covers" data-attributes="{&quot;cover_type&quot;:&quot;take-action&quot;,&quot;initialRowsLimit&quot;:1,&quot;title&quot;:&quot;Take Action Cover Block&quot;,&quot;description&quot;:&quot;&quot;,&quot;tags&quot;:[],&quot;post_types&quot;:[],&quot;posts&quot;:[],&quot;version&quot;:2,&quot;layout&quot;:&quot;grid&quot;,&quot;isExample&quot;:false,&quot;readMoreText&quot;:&quot;Load more&quot;,&quot;covers&quot;:[],&quot;className&quot;:&quot;is-style-take-action&quot;}"></div></div>
> <!-- /wp:planet4-blocks/covers -->
> 
> <!-- wp:planet4-blocks/covers {"cover_type":"campaign","title":"Campaign Cover Block","tags":[8],"className":"is-style-campaign"} -->
> <div class="wp-block-planet4-blocks-covers is-style-campaign"><div data-hydrate="planet4-blocks/covers" data-attributes="{&quot;cover_type&quot;:&quot;campaign&quot;,&quot;initialRowsLimit&quot;:1,&quot;title&quot;:&quot;Campaign Cover Block&quot;,&quot;description&quot;:&quot;&quot;,&quot;tags&quot;:[8],&quot;post_types&quot;:[],&quot;posts&quot;:[],&quot;version&quot;:2,&quot;layout&quot;:&quot;grid&quot;,&quot;isExample&quot;:false,&quot;readMoreText&quot;:&quot;Load more&quot;,&quot;covers&quot;:[],&quot;className&quot;:&quot;is-style-campaign&quot;}"></div></div>
> <!-- /wp:planet4-blocks/covers -->
> 
> <!-- wp:planet4-blocks/covers {"cover_type":"campaign","title":"Campaign Cover Block 2","description":"Description of the campaign cover 2\u003cbr\u003eLayout: carousel\u003cbr\u003eTags: consumption + renewables","tags":[8,7],"layout":"carousel","className":"is-style-campaign"} -->
> <div class="wp-block-planet4-blocks-covers is-style-campaign"><div data-hydrate="planet4-blocks/covers" data-attributes="{&quot;cover_type&quot;:&quot;campaign&quot;,&quot;initialRowsLimit&quot;:1,&quot;title&quot;:&quot;Campaign Cover Block 2&quot;,&quot;description&quot;:&quot;Description of the campaign cover 2&lt;br&gt;Layout: carousel&lt;br&gt;Tags: consumption + renewables&quot;,&quot;tags&quot;:[8,7],&quot;post_types&quot;:[],&quot;posts&quot;:[],&quot;version&quot;:2,&quot;layout&quot;:&quot;carousel&quot;,&quot;isExample&quot;:false,&quot;readMoreText&quot;:&quot;Load more&quot;,&quot;covers&quot;:[],&quot;className&quot;:&quot;is-style-campaign&quot;}"></div></div>
> <!-- /wp:planet4-blocks/covers -->
> 
> <!-- wp:planet4-blocks/covers {"cover_type":"campaign","initialRowsLimit":2,"title":"Campaign Cover Block 3","description":"Description of the campaign cover 2\u003cbr\u003eLayout: grid 2 rows\u003cbr\u003eTags: consumption + renewables + oceans + climate","tags":[8,7,6,26],"className":"is-style-campaign"} -->
> <div class="wp-block-planet4-blocks-covers is-style-campaign"><div data-hydrate="planet4-blocks/covers" data-attributes="{&quot;cover_type&quot;:&quot;campaign&quot;,&quot;initialRowsLimit&quot;:2,&quot;title&quot;:&quot;Campaign Cover Block 3&quot;,&quot;description&quot;:&quot;Description of the campaign cover 2&lt;br&gt;Layout: grid 2 rows&lt;br&gt;Tags: consumption + renewables + oceans + climate&quot;,&quot;tags&quot;:[8,7,6,26],&quot;post_types&quot;:[],&quot;posts&quot;:[],&quot;version&quot;:2,&quot;layout&quot;:&quot;grid&quot;,&quot;isExample&quot;:false,&quot;readMoreText&quot;:&quot;Load more&quot;,&quot;covers&quot;:[],&quot;className&quot;:&quot;is-style-campaign&quot;}"></div></div>
> <!-- /wp:planet4-blocks/covers -->
> 
> <!-- wp:planet4-blocks/covers {"title":"Content Cover Block 1","description":"Layout: carousel\u003cbr\u003eTags: consumption, renewables\u003cbr\u003ePost type: press release","tags":[8,7],"post_types":[14],"layout":"carousel","className":"is-style-content"} -->
> <div class="wp-block-planet4-blocks-covers is-style-content"><div data-hydrate="planet4-blocks/covers" data-attributes="{&quot;cover_type&quot;:&quot;content&quot;,&quot;initialRowsLimit&quot;:1,&quot;title&quot;:&quot;Content Cover Block 1&quot;,&quot;description&quot;:&quot;Layout: carousel&lt;br&gt;Tags: consumption, renewables&lt;br&gt;Post type: press release&quot;,&quot;tags&quot;:[8,7],&quot;post_types&quot;:[14],&quot;posts&quot;:[],&quot;version&quot;:2,&quot;layout&quot;:&quot;carousel&quot;,&quot;isExample&quot;:false,&quot;readMoreText&quot;:&quot;Load more&quot;,&quot;covers&quot;:[],&quot;className&quot;:&quot;is-style-content&quot;}"></div></div>
> <!-- /wp:planet4-blocks/covers -->
> 
> <!-- wp:planet4-blocks/covers {"title":"Content Cover Block 2","description":"Layout: grid\u003cbr\u003eTags: -\u003cbr\u003ePost type: story, press release","post_types":[16,14],"className":"is-style-content"} -->
> <div class="wp-block-planet4-blocks-covers is-style-content"><div data-hydrate="planet4-blocks/covers" data-attributes="{&quot;cover_type&quot;:&quot;content&quot;,&quot;initialRowsLimit&quot;:1,&quot;title&quot;:&quot;Content Cover Block 2&quot;,&quot;description&quot;:&quot;Layout: grid&lt;br&gt;Tags: -&lt;br&gt;Post type: story, press release&quot;,&quot;tags&quot;:[],&quot;post_types&quot;:[16,14],&quot;posts&quot;:[],&quot;version&quot;:2,&quot;layout&quot;:&quot;grid&quot;,&quot;isExample&quot;:false,&quot;readMoreText&quot;:&quot;Load more&quot;,&quot;covers&quot;:[],&quot;className&quot;:&quot;is-style-content&quot;}"></div></div>
> <!-- /wp:planet4-blocks/covers -->
> 
> <!-- wp:planet4-blocks/covers {"title":"Content Cover Block 3","description":"Layout: grid\u003cbr\u003eTags: consumption, renewables\u003cbr\u003ePost type: -","tags":[8,7],"className":"is-style-content"} -->
> <div class="wp-block-planet4-blocks-covers is-style-content"><div data-hydrate="planet4-blocks/covers" data-attributes="{&quot;cover_type&quot;:&quot;content&quot;,&quot;initialRowsLimit&quot;:1,&quot;title&quot;:&quot;Content Cover Block 3&quot;,&quot;description&quot;:&quot;Layout: grid&lt;br&gt;Tags: consumption, renewables&lt;br&gt;Post type: -&quot;,&quot;tags&quot;:[8,7],&quot;post_types&quot;:[],&quot;posts&quot;:[],&quot;version&quot;:2,&quot;layout&quot;:&quot;grid&quot;,&quot;isExample&quot;:false,&quot;readMoreText&quot;:&quot;Load more&quot;,&quot;covers&quot;:[],&quot;className&quot;:&quot;is-style-content&quot;}"></div></div>
> <!-- /wp:planet4-blocks/covers -->
> ```

Hi @mardelnet , Thanks for spotting the slash escape issue. I've updated the PR and also fixed all the other issues. Please have a look.

mleray commented on src/Migrations/M032MigrateCampaignCoversToP4ColumnsBlock.php:
```diff
@@ -0,0 +1,165 @@
+                echo 'Migrating post ', $post->ID, "\n"; // phpcs:ignore
+
+                $post_update = array(
+                    'ID' => $post->ID,
```
Why not use the `$current_post_id` variable here?

mleray commented on src/Migrations/M032MigrateCampaignCoversToP4ColumnsBlock.php:
```diff
@@ -0,0 +1,165 @@
+                $result = wp_update_post(wp_slash($post_update));
+
+                if ($result === 0) {
+                    throw new \Exception("There was an error trying to update the post #" . $post->ID);
```
Same question here, we could use `$current_post_id`

sagarsdeshmukh replied:
It was a miss, Thanks heads up 🙏

mleray commented:
@sagarsdeshmukh [the Covers block has been moved to the theme](https://github.com/greenpeace/planet4-master-theme/pull/2402), maybe as part of this migration PR we could also remove the Campaign style from the style picker? We can remove the rest of the related code in a follow-up PR since I expect it will be a bit tricky.

sagarsdeshmukh commented:
> @sagarsdeshmukh [the Covers block has been moved to the theme](https://github.com/greenpeace/planet4-master-theme/pull/2402), maybe as part of this migration PR we could also remove the Campaign style from the style picker? We can remove the rest of the related code in a follow-up PR since I expect it will be a bit tricky.

Hi @mleray , Thanks for your feedback. This PR is base with `migrations` branch and it won't have the merged covers block changes, hence I opened a separate [PR](https://github.com/greenpeace/planet4-master-theme/pull/2407) in which we remove the covers campaign style.

mardelnet commented:
@sagarsdeshmukh 
1. I can't test this PR locally. When I try to add Covers block I get the error: "Your site doesn’t include support for the "planet4-blocks/covers" block. You can leave this block intact, convert its content to a Custom HTML block, or remove it entirely."
2. The [test environment](https://www-dev.greenpeace.org/test-mars/) for this PR is no longer working either.

sagarsdeshmukh commented:
> @sagarsdeshmukh
> 
> 1. I can't test this PR locally. When I try to add Covers block I get the error: "Your site doesn’t include support for the "planet4-blocks/covers" block. You can leave this block intact, convert its content to a Custom HTML block, or remove it entirely."
> 2. The [test environment](https://www-dev.greenpeace.org/test-mars/) for this PR is no longer working either.

Hi @mardelnet , The issue occurred because the base branch migration was 8 commits behind the main branch, and the covers block was moved to the master theme and removed from the blocks plugin. I've now rebased my branch. Could you please check it now?

mardelnet reviewed (commented):
@sagarsdeshmukh, also remember to remove the block code and the associated tests.

sagarsdeshmukh commented:
> @sagarsdeshmukh, also remember to remove the block code and the associated tests.

Fixed!

mardelnet reviewed (commented):
Hey @sagarsdeshmukh 
I still see many references to the Campaign Covers in the codebase, should we remove them as well? 
Just to mention a few of them (there are more):
- src/Blocks/Covers.php
- assets/src/blocks/Covers/CampaignCovers.js
- assets/src/blocks/Covers/CoversEditor.js
- assets/src/blocks/Covers/imageSizes.js

mardelnet commented on src/Migrations/Utils/Functions.php:
```diff
@@ -14,13 +15,17 @@ class Functions
      */
-    public static function get_posts_using_specific_block(string $block_name, array $post_types): mixed
+    //phpcs:ignore Generic.Files.LineLength.MaxExceeded
+    public static function get_posts_using_specific_block(string $block_name, array $post_types, array $post_status): mixed
```
I think we have to make the `$post_status` parameter optional.
Otherwise, there will be issues with other parts of the code using this function (check the `src/Migrations/M031MigrateMediaBlockToAudioVideoBlock.php` file) and the migrator will crash.

mardelnet replied:
@sagarsdeshmukh you can check my implementation for this here: https://github.com/greenpeace/planet4-master-theme/blob/cf8b234ce4490117345c6506016ccec6afd69ee6/src/Migrations/Utils/Functions.php#L23

mardelnet reviewed (commented):
It seems that once [this bug](https://github.com/greenpeace/planet4-master-theme/pull/2349#discussion_r1814391953) is fixed and the migration is completed, the Content Covers are broken:

![Edit-Page-“pedro”-‹-Greenpeace-—-WordPress](https://github.com/user-attachments/assets/803c1b1b-7960-4304-9eea-5f5d626681a3)

mardelnet commented:
Closing this PR in favor of https://github.com/greenpeace/planet4-master-theme/pull/2435
